### PR TITLE
migrate src to TS

### PR DESCRIPTION
**Pull request by longlho**
- Migrate all src/ to TypeScript. There will be a follow up diff to
improve types since there're a lot of `any` casting right now.
- Reduce bundle size by removing `prop-types` dependency.
- BREAKING CHANGE: Remove `intlShape` from top level export. `IntlShape`
interface should replace that.
- Reduce bundle size by removing `react-is`. This is the result of
changing `withIntl` API to a strict `withIntl(Component, opts)` instead
of the previous `componentOrOptions` 1st param introduced in #1211.
Makes typechecking more straightforward.
- Fix tests & remove redundant tests.
- Follow up diff will include documentation updates & more work on types

### CODE REVIEW

**Comment by redonkulus** [comment]
Removing intlShape and prop-types is a pretty big change. If projects are not using TS then they will lose the prop type checking. I would prefer to support both.

**Comment by longlho** [comment]
So the reasons i rm `intlShapre` and `prop-types` are:
- It’s only useful in dev mode. Prod build removes `prop-types`
- Most IDEs actually pick up type def even for regular js and would show error 
- Popular libs that convert to TS removes prop-types as well
- Maintenance overhead

File: src/components/message.tsx [hunk 1]
```diff
@@ -66,10 +67,7 @@ class FormattedMessage extends React.Component {
       values,
     };
 
-    return (
-      !shallowEquals(this.props, nextPropsToCheck) ||
-      !shallowEquals(this.state, nextState)
-    );
+    return !shallowEquals(this.props, nextPropsToCheck);
```
[the remark this answers is not in the document]
Comment (longlho) [reply src/components/message.tsx]: cause `message` is a stateless component :)

File: package.json [hunk 2]
```diff
@@ -62,21 +62,18 @@
...
+    "@types/react": "^16.8.20",
     "hoist-non-react-statics": "^3.3.0",
-    "intl-format-cache": "^3.0.0",
+    "intl-format-cache": "^3.0.2",
     "intl-locales-supported": "^1.0.10",
     "intl-messageformat": "^4.0.0",
     "intl-relativeformat": "^6.1.0",
     "invariant": "^2.1.1",
-    "react-is": "^16.3.1",
+    "react": "^16.8.0",
```
[the remark this answers is not in the document]
Comment (longlho) [reply package.json]: yup I'll revert that

File: src/components/message.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (longlho) [reply src/components/message.tsx]: cause `message` is a stateless component :)

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (longlho) [reply package.json]: yup I'll revert that

**Comment by longlho** [comment]
`-2K` minified (5%), `-5K` unminified

**Comment by redonkulus** [comment]
Can you update the upgrade guide about the removal of IntlShape?

**Comment by longlho** [comment]
done

**Pull request merged**